### PR TITLE
feat: add rate limiter implementation

### PR DESCRIPTION
**Pull request by h2non**
Provides a very simple rate limiter middleware based on rate per second and time window.

Additionally it allows you to filter and ignore certain traffic via `Filter` functions and ability to customize how to reply to the client in case of rate limit reached.

Tests should be improved covering better scenarios.

### CODE REVIEW

File: ratelimit.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	mw.UsePriority("request", layer.TopHead, l.LimitHTTP)
+}
+
+// LimitHTTP limits an incoming HTTP request.
+// If some filter passes, the request won't be limited.
+// This method is used internally, but made public for public testing.
+func (l *Limiter) LimitHTTP(h http.Handler) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// Pass filters
+		for _, filter := range l.filters {
```
Comment (groyoh) [inline ratelimit.go]: Shouldn't filter work the other way around? When I filter for GET requests, I would expect only these requests to be limited.

File: ratelimit.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	l.lm.Unlock()
+}
+
+// capacity is used to read the current bucket capacity.
+func (l *Limiter) capacity() int {
+	return int(l.bucket.Capacity())
+}
+
+// remaining is used to read the current pending remaining available buckets.
+func (l *Limiter) reamining() int {
```
Comment (groyoh) [inline ratelimit.go]: Typo?

File: ratelimit.go (hunk 1 above)
Comment (h2non) [inline ratelimit.go]: I agree and also had doubts about the semantic here. We could invert the behavior or add other semantic for whitelist, such as: exceptions.

File: ratelimit.go (hunk 2 above)
Comment (h2non) [inline ratelimit.go]: Fixed.

File: ratelimit.go (hunk 1 above)
Comment (h2non) [inline ratelimit.go]: I've just changed filters behavior. Now filters are designed to determine when should apply the rate limiter. All the filters must pass in order to apply the limiter. No filters means all the traffic will be limited.

From the other hand, I've supported exceptions as well, which allows the developer to determine in opposite scenario: I want to do not limit always certain typo of requests.

Take a look and let's move forward with this.

File: ratelimit.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+// SetResponder sets a custom function to reply in case of rate limit reached.
+func (l *Limiter) SetResponder(fn http.HandlerFunc) {
+	l.responder = fn
+}
+
+// AddFilter adds a new rate limiter whitelist filter.
+// If the filter matches, the traffic won't be limited.
+func (l *Limiter) AddFilter(fn ...Filter) {
```
Comment (groyoh) [inline ratelimit.go]: I think we should try and stay consistent across package. I think in other packages we use `Filter`.

File: ratelimit.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	l.filters = append(l.filters, fn...)
+}
+
+// AddException adds a new rate limiter whitelist filter.
+// If the filter matches, the traffic won't be limited.
```
Comment (groyoh) [inline ratelimit.go]: The doc doesn't match.

File: ratelimit.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+// capacity is used to read the current bucket capacity.
+func (l *Limiter) capacity() int {
+	return int(l.bucket.Capacity())
+}
+
+// remaining is used to read the current pending remaining available buckets.
+func (l *Limiter) remaining() int {
+	if remaining := int(l.bucket.Available()); remaining > 0 {
```
Comment (groyoh) [inline ratelimit.go:133]: Can it actually be negative?

Comment (h2non) [inline ratelimit.go:133]: Yes. It could be.

File: ratelimit.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	filters []Filter
+	// exceptions stores a list of exceptions to determine if should not apply the rate limiter.
+	exceptions []Exception
+}
+
+// NewTimeLimiter creates a new time based rate limiter middleware.
+func NewTimeLimiter(timeWindow time.Duration, capacity int64) *Limiter {
+	return &Limiter{
+		responder:     RateLimitResponder,
+		timeWindow:    timeWindow,
```
Comment (groyoh) [inline ratelimit.go]: What's the point of saving the `timeWindow`?

File: ratelimit.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+		l.trackTime()
+		h.ServeHTTP(w, r)
+	}
+}
+
+// resetTime is used to calculate the pending reset time to wait.
+func (l *Limiter) resetTime() time.Duration {
+	l.lm.Lock()
+	defer l.lm.Unlock()
+	return time.Now().Sub(l.lastAvailable) / time.Second
```
Comment (groyoh) [inline ratelimit.go]: I don't get `time.Now().Sub(l.lastAvailable) / time.Second` is the reset time. Could you maybe detail that? It doesn't really make sense to me.

File: ratelimit.go (hunk 3 above)
Comment (h2non) [inline ratelimit.go]: I agree. Will change it.

File: ratelimit.go (hunk 6 above)
Comment (h2non) [inline ratelimit.go]: Fixed.

File: ratelimit.go (hunk 7 above)
Comment (h2non) [inline ratelimit.go]: Well, the idea is basically to calculate the pending wait time until the bucket has available tokens. 
My initial approach is not good, actually. I'll rethink about this.

File: ratelimit.go (hunk 6 above)
Comment (h2non) [inline ratelimit.go]: I used it before, then I did some refactors and the field is still there. I'll remove it after some improvements, if it's really do not used.

**Comment by h2non** [comment]
I decided to simplify this. Ideally `ratelimit` package should provide the pending amount of time the client has to wait until some bucket is available. That would simplify a lot the things here.

Since it's a nice to have feature for now, let's keep as pending feature request.
For sure, if you want you can go ahead with it.

**Comment by h2non** [comment]
@groyoh Please, confirm this to move forward. The missing feature has been created here: #5

**Comment by groyoh** [comment]
Yeah looks good :+1:

**Pull request merged**